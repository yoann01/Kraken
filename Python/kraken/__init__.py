"""Kraken Framework."""

import logging
import os

__all__ = ['core', 'helpers', 'plugins', 'ui']


krakenPath = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..'))
if os.environ.get('KRAKEN_PATH', None) is None:
    os.environ['KRAKEN_PATH'] = krakenPath

krakenExtsPath = os.path.join(krakenPath, 'Exts')
if 'FABRIC_EXTS_PATH' in os.environ:
    if krakenExtsPath not in os.environ['FABRIC_EXTS_PATH']:
        os.environ['FABRIC_EXTS_PATH'] = krakenExtsPath + os.pathsep + os.environ['FABRIC_EXTS_PATH']

krakenPresetsPath = os.path.join(krakenPath, 'Presets', 'DFG')
if 'FABRIC_DFG_PATH' in os.environ:
    if krakenPresetsPath not in os.environ['FABRIC_DFG_PATH']:
        os.environ['FABRIC_DFG_PATH'] = krakenPresetsPath + os.pathsep + os.environ['FABRIC_DFG_PATH']


logging.basicConfig(format='[KRAKEN] %(levelname)s: %(message)s', level=logging.INFO)

# Custom inform level for use with UI label getting added to the status bar.
logging.INFORM = 25
logging.addLevelName(logging.INFORM, 'INFORM')
logging.Logger.inform = lambda inst, msg, *args, **kwargs: inst.log(logging.INFORM, msg, *args, **kwargs)


def release():
    print "RELEASE THE KRAKEN!!!!"
