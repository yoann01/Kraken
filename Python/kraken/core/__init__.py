"""Kraken Core."""

VERSION_MAJOR = 1
VERSION_MINOR = 2
VERSION_BUILD = 6
VERSION_SUFFIX = ""


def getVersion():
    """Contatenates the version globals and returns the current version of
    Kraken.

    Returns:
        str: Current version of Kraken.

    """

    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString
