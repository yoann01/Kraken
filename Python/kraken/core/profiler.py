"""Kraken - core.profiler module.

Classes:
Profiler - Profiler Object.

"""

import time
import operator


class _ProfilerItem(object):

    def __init__(self, label):
        super(_ProfilerItem, self).__init__()

        t = time.time()
        self.label = label
        self.start = t
        self.end = t
        self.children = []


    def addChild(self, item):
        self.children.append(item)


    def endProfiling(self):
        self.end = time.time()


class Profiler(object):
    """Kraken profiler object for debugging performance issues."""

    __instance = None


    def __init__(self):
        super(Profiler, self).__init__()
        self.reset()


    def reset(self):
        """Resets the profiler for generating a new report"""

        self.__roots = []
        self.__stack = []


    def push(self, label):

        """Adds a new child to the profiling tree and activates it.

        Args:
            label (str): The label of the next level of the stack.

        """

        item = _ProfilerItem(label)
        if len(self.__stack) == 0:
            self.__roots.append(item)
        else:
            self.__stack[-1].addChild(item)

        self.__stack.append(item)


    def pop(self):
        """Deactivates the current item in the tree and returns the profiler to
        the parent item"""

        if len(self.__stack) == 0:
            raise Exception("""Unable to close bracket. Pop has been called more """ +
                            """times than push.""")

        self.__stack[-1].endProfiling()
        self.__stack.pop()


    def generateReport(self, listFunctionTotals=False):
        """Returns a report string containing all the data gathered turing
        profiling.

        Args:
            listFunctionTotals (bool): list information relating to the total time spent in each function.

        Returns:
            str: The profiler report.

        """

        if len(self.__stack) != 0:
            raise Exception("""Profiler brackets not closed properly. """ +
                            """Pop must be called for every call to push. Pop """ +
                            """needs to be called another """ +
                            str(len(self.__stack)) + """ times""")

        report = []
        report.append("--callstack--")

        functions = {}

        def reportItem(item, indent):
            duration = item.end - item.start
            report.append(indent + item.label + ' duration: ' + str(duration))
            if item.label not in functions:
                functions[item.label] = duration
            else:
                functions[item.label] += duration

            for childItem in item.children:
                reportItem(childItem, indent + '  ')

        for rootItem in self.__roots:
            reportItem(rootItem, '  ')

        if listFunctionTotals:
            report.append("--functions--")

            def reverse_numeric(x, y):
                return y < x

            sorted_fns = sorted(functions.items(), key=operator.itemgetter(1),
                                cmp=reverse_numeric)

            for fn_tuple in sorted_fns:
                report.append(str(fn_tuple[1]) + ': ' + fn_tuple[0])

        return '\n'.join(report)


    @classmethod
    def getInstance(cls):
        """This class method returns the singleton instance for the Profiler

        Returns:
            object: The singleton profiler instance.

        """

        if cls.__instance is None:
            cls.__instance = Profiler()

        return cls.__instance
