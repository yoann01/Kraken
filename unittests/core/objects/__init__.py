
import unittest

from attributes import suite as attributeSuite
from components import suite as componentSuite
from constraints import suite as constraintSuite
from operators import suite as operatorSuite

from test_component_group import suite as componentGroupSuite
from test_container import suite as containerSuite
from test_control import suite as controlSuite
from test_ctrlSpace import suite as ctrlSpaceSuite
from test_curve import suite as curveSuite
from test_hierarchy_group import suite as hierarchyGroupSuite
from test_joint import suite as jointSuite
from test_layer import suite as layerSuite
from test_locator import suite as locatorSuite
from test_object_3d import suite as object3DSuite
from test_rig import suite as rigSuite
from test_scene_item import suite as sceneItemSuite
from test_transform import suite as transformSuite

loadAttributeSuite = attributeSuite()
loadComponentSuite = componentSuite()
loadConstraintSuite = constraintSuite()
loadOperatorSuite = operatorSuite()

loadComponentGroupSuite = componentGroupSuite()
loadContainerSuite = containerSuite()
loadControlSuite = controlSuite()
loadCtrlSpaceSuite = ctrlSpaceSuite()
loadCurveSuite = curveSuite()
loadHierarchyGroupSuite = hierarchyGroupSuite()
loadJointSuite = jointSuite()
loadLayerSuite = layerSuite()
loadLocatorSuite = locatorSuite()
loadObject3DSuite = object3DSuite()
loadRigSuite = rigSuite()
loadSceneItemSuite = sceneItemSuite()
loadTransformSuite = transformSuite()


def suite():
    suites = [
        loadAttributeSuite,
        loadComponentSuite,
        loadConstraintSuite,
        loadOperatorSuite,
        loadContainerSuite,
        loadControlSuite,
        loadCtrlSpaceSuite,
        loadCurveSuite,
        loadHierarchyGroupSuite,
        loadJointSuite,
        loadLayerSuite,
        loadLocatorSuite,
        loadObject3DSuite,
        loadRigSuite,
        loadSceneItemSuite,
        loadTransformSuite]

    return unittest.TestSuite(suites)


if __name__ == '__main__':
    runner = unittest.TextTestRunner()
    runner.run(suite())
