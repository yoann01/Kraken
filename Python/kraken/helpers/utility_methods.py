from kraken.core.maths.math_object import MathObject
from kraken.core.maths.vec3 import Vec3
from kraken.core.maths.xfo import Xfo
from kraken.core.maths.quat import Quat

from kraken.core.maths import decodeValue


def logHierarchy(kObject):
    """Traverses the given Kraken hierarchy and logs the names of all the objects.

    Args:
        Object: The object to start logging on.

    """

    print kObject.getDecoratedPath()
    for i in xrange(kObject.getNumChildren()):
        child = kObject.getChildByIndex(i)
        logHierarchy(child)


def __convertFromJSON(jsonData):

    if type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertFromJSON(item))
        return newList
    elif type(jsonData) is dict:
        if '__mathObjectClass__' in jsonData.keys():
            return decodeValue(jsonData)
        for key, value in jsonData.iteritems():
            jsonData[key] = __convertFromJSON(value)
    return jsonData


def prepareToLoad(jsonData):
    """Prepares the json data for loading into kraken.

    Args:
        jsonData (dict): The JSON data to be prepared.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __convertFromJSON(jsonData)


def __convertToJSON(jsonData):

    if isinstance(jsonData, MathObject):
        return jsonData.jsonEncode()
    elif type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertToJSON(item))
        return newList
    elif type(jsonData) is dict:
        newDict = {}
        for key, value in jsonData.iteritems():
            newDict[key] = __convertToJSON(value)
        return newDict
    return jsonData


def prepareToSave(jsonData):
    """Prepares the json data for serialization.

    Args:
        jsonData (dict): The JSON data to be prepared.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __convertToJSON(jsonData)


def __mirrorData(jsonData, plane):

    if isinstance(jsonData, Vec3):
        newVec3 = Vec3(jsonData)
        if plane == 0:
            newVec3.x = -newVec3.x
        elif plane == 1:
            newVec3.y = -newVec3.y
        elif plane == 2:
            newVec3.z = -newVec3.z
        return newVec3

    if isinstance(jsonData, Quat):
        newQuat = Quat(jsonData)
        newQuat.mirror(plane)
        return newQuat

    elif isinstance(jsonData, Xfo):
        newXfo = Xfo(jsonData)
        if plane == 0:
            newXfo.tr.x = -newXfo.tr.x
        elif plane == 1:
            newXfo.tr.y = -newXfo.tr.y
        elif plane == 2:
            newXfo.tr.z = -newXfo.tr.z

        newXfo.ori.mirror(plane)
        return newXfo

    elif type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__mirrorData(item, plane))
        return newList

    elif type(jsonData) is dict:
        newDict = {}
        for key, value in jsonData.iteritems():
            newDict[key] = __mirrorData(value, plane)
        return newDict

    return jsonData


def mirrorData(jsonData, plane):
    """Prepares the json data for serialization.

    Args:
        jsonData (dict): The JSON data to be prepared.
        plane (int): The plane to mirror across.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __mirrorData(jsonData, plane)
