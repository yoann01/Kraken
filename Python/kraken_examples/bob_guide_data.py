"""Example data that can be loaded onto a guide rig."""

from kraken.core.maths import Vec2, Vec3, Euler, Quat, Xfo


bob_guide_data = {
    "components":[
        {
            "class": "kraken_components.biped.spine_component.SpineComponentGuide",
            'cogPosition': Vec3(0.0, 11.1351, -0.1382),
            'spine01Position': Vec3(0.0, 11.1351, -0.1382),
            'spine02Position': Vec3(0.0, 11.8013, -0.1995),
            'spine03Position': Vec3(0.0, 12.4496, -0.3649),
            'spine04Position': Vec3(0.0, 13.1051, -0.4821),
            'numDeformers': 6
        },
        {
            "class": "kraken_components.biped.neck_component.NeckComponentGuide",
            "neckPosition": Vec3(0.0, 16.5572, -0.6915),
            "neckEndPosition": Vec3(0.0, 17.4756, -0.421)
        },
        {
            "class": "kraken_components.biped.head_component.HeadComponentGuide",
            "headPosition": Vec3(0.0, 17.4756, -0.421),
            "headEndPosition": Vec3(0.0, 19.5, -0.421),
            "eyeLeftPosition": Vec3(0.3497, 18.0878, 0.6088),
            "eyeRightPosition": Vec3(-0.3497, 18.0878, 0.6088),
            "jawPosition": Vec3(0.0, 17.613, -0.2731)
        },
        {
            "class": "kraken_components.biped.clavicle_component.ClavicleComponentGuide",
            "name": "Clavicle",
            "location": "L",
            "clavicleXfo": Xfo(Vec3(0.1322, 15.403, -0.5723)),
            "clavicleUpVXfo": Xfo(Vec3(0.0, 1.0, 0.0)),
            "clavicleEndXfo": Xfo(Vec3(2.27, 15.295, -0.753))
        },
        {
            "class": "kraken_components.biped.clavicle_component.ClavicleComponentGuide",
            "name": "Clavicle",
            "location": "R",
            "clavicleXfo": Xfo(Vec3(-0.1322, 15.403, -0.5723)),
            "clavicleUpVXfo": Xfo(Vec3(0.0, 1.0, 0.0)),
            "clavicleEndXfo": Xfo(Vec3(-2.27, 15.295, -0.753))
        },
        {
            "class": "kraken_components.biped.arm_component.ArmComponentGuide",
            "name": "Arm",
            "location": "L",
            "bicepXfo": Xfo(Vec3(2.27, 15.295, -0.753)),
            "forearmXfo": Xfo(Vec3(5.039, 13.56, -0.859)),
            "wristXfo": Xfo(Vec3(7.1886, 12.2819, 0.4906)),
            "handXfo": Xfo(tr=Vec3(7.1886, 12.2819, 0.4906),
                           ori=Quat(Vec3(-0.0865, -0.2301, -0.2623), 0.9331)),
            "bicepFKCtrlSize": 1.75,
            "forearmFKCtrlSize": 1.5
        },
        {
            "class": "kraken_components.biped.arm_component.ArmComponentGuide",
            "name": "Arm",
            "location": "R",
            "bicepXfo": Xfo(Vec3(-2.27, 15.295, -0.753)),
            "forearmXfo": Xfo(Vec3(-5.039, 13.56, -0.859)),
            "wristXfo": Xfo(Vec3(-7.1886, 12.2819, 0.4906)),
            "handXfo": Xfo(tr=Vec3(-7.1886, 12.2819, 0.4906),
                           ori=Quat(Vec3(-0.2301, -0.0865, -0.9331), 0.2623)),
            "bicepFKCtrlSize": 1.75,
            "forearmFKCtrlSize": 1.5
        },
        {
            "class": "kraken_components.biped.leg_component.LegComponentGuide",
            "name": "Leg",
            "location": "L",
            "femurXfo": Xfo(Vec3(0.9811, 9.769, -0.4572)),
            "kneeXfo": Xfo(Vec3(1.4488, 5.4418, -0.5348)),
            "ankleXfo": Xfo(Vec3(1.841, 1.1516, -1.237)),
            "toeXfo": Xfo(Vec3(1.85, 0.4, 0.25)),
            "toeTipXfo": Xfo(Vec3(1.85, 0.4, 1.5))
        },
        {
            "class": "kraken_components.biped.leg_component.LegComponentGuide",
            "name": "Leg",
            "location": "R",
            "femurXfo": Xfo(Vec3(-0.9811, 9.769, -0.4572)),
            "kneeXfo": Xfo(Vec3(-1.4488, 5.4418, -0.5348)),
            "ankleXfo": Xfo(Vec3(-1.841, 1.1516, -1.237)),
            "toeXfo": Xfo(Vec3(-1.85, 0.4, 0.25)),
            "toeTipXfo": Xfo(Vec3(-1.85, 0.4, 1.5))
        }
    ],
    "connections": [
        {
            "_comment": "Neck to Spine",
            "source": "spine:M.spineEnd",
            "target": "neck:M.neckBase"
        },
        {
            "_comment": "Head to Neck",
            "source": "neck:M.neckEnd",
            "target": "head:M.headBase"
        },
        {
            "_comment": "LClavicle to Spine",
            "source": "spine:M.spineEnd",
            "target": "Clavicle:L.spineEnd"
        },
        {
            "_comment": "LArm to LClavicle",
            "source": "Clavicle:L.clavicleEnd",
            "target": "Arm:L.root"
        },
        {
            "_comment": "RArm to RClavicle",
            "source": "Clavicle:R.clavicleEnd",
            "target": "Arm:R.root"
        },
        {
            "_comment": "RClavicle to Spine",
            "source": "spine:M.spineEnd",
            "target": "Clavicle:R.spineEnd"
        },
        {
            "_comment": "LLeg To Pelvis Connections",
            "source": "spine:M.pelvis",
            "target": "Leg:L.pelvisInput"
        },
        {
            "_comment": "RLeg To Pelvis Connections",
            "source": "spine:M.pelvis",
            "target": "Leg:R.pelvisInput"
        }
    ]
}
