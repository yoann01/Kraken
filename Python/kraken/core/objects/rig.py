"""Kraken - objects.rig module.

Classes:
Rig -- Rig representation.

"""

import importlib
import json
import os

from container import Container
from kraken.core.kraken_system import KrakenSystem
from kraken.core.profiler import Profiler
from kraken.helpers.utility_methods import prepareToSave, prepareToLoad


class Rig(Container):
    """Rig object."""

    def __init__(self, name='rig', metaData=None):
        super(Rig, self).__init__(name, metaData=metaData)

    # ====================
    # Load / Save Methods
    # ====================
    def writeRigDefinitionFile(self, filepath):
        """Load a rig definition from a file on disk.

        Args:
            filepath (str): The file path of the rig definition file.

        Returns:
            bool: True if successful.

        """

        Profiler.getInstance().push("writeRigDefinitionFile:" + filepath)

        jsonData = self.getData()

        # now preprocess the data ready for saving to disk.
        pureJSON = prepareToSave(jsonData)

        with open(filepath,'w') as rigFile:
            rigFile.write(json.dumps(pureJSON, indent=2))

        Profiler.getInstance().pop()

    def loadRigDefinitionFile(self, filepath):
        """Load a rig definition from a file on disk.

        Args:
            filepath (str): The file path of the rig definition file.

        Returns:
            bool: True if successful.

        """

        Profiler.getInstance().push("LoadRigDefinitionFile:" + filepath)

        if not os.path.exists(filepath):
            raise Exception("File not found:" + filepath)

        with open(filepath) as rigFile:
            jsonData = json.load(rigFile)

        # now preprocess the data ready for loading.
        jsonData = prepareToLoad(jsonData)

        self.loadRigDefinition(jsonData)
        Profiler.getInstance().pop()

    def _loadComponents(self, componentsJson):
        """Loads components from a JSON dict.

        Args:
            componentsJson (dict): Dictionary of components to load.

        """


        Profiler.getInstance().push("__loadComponents")

        krakenSystem = KrakenSystem.getInstance()

        for componentData in componentsJson:
            # trim off the class name to get the module path.
            modulePath = '.'.join(componentData['class'].split('.')[:-1])

            if modulePath is not "":
                try:
                    importlib.import_module(modulePath)
                except:
                    print "Warning: Error finding module path: " + modulePath
                    continue

            componentClass = krakenSystem.getComponentClass(componentData['class'])
            if 'name' in componentData:
                component = componentClass(name=componentData['name'], parent=self)
            else:
                component = componentClass(parent=self)
            component.loadData(componentData)

        Profiler.getInstance().pop()

    def _makeConnections(self, connectionsJson):
        """Makes connections based on JSON dict.

        Args:
            connectionsJson (dict): Dictionary of connections to make.

        """

        Profiler.getInstance().push("__makeConnections")

        for connectionData in connectionsJson:

            sourceComponentDecoratedName, outputName = connectionData['source'].split('.')
            targetComponentDecoratedName, inputName = connectionData['target'].split('.')

            connectionFailure = False

            sourceComponent = self.getChildByDecoratedName(sourceComponentDecoratedName)
            if sourceComponent is None:
                print("Warning: Error making connection:" + connectionData['source'] + " -> " +
                      connectionData['target'] + ". Source component not found:" + sourceComponentDecoratedName)
                connectionFailure = True

            targetComponent = self.getChildByDecoratedName(targetComponentDecoratedName)
            if targetComponent is None:
                print("Warning: Error making connection:" + connectionData['source'] + " -> " +
                      connectionData['target'] + ". Target component not found:" + targetComponentDecoratedName)
                connectionFailure = True

            outputPort = sourceComponent.getOutputByName(outputName)
            if outputPort is None:
                print("Warning: Error making connection:" + connectionData['source'] + " -> " +
                      connectionData['target'] + ". Output '" + outputName + "' not found on Component:" + sourceComponent.getPath())
                connectionFailure = True

            inputPort = targetComponent.getInputByName(inputName)
            if inputPort is None:
                print("Warning: Error making connection:" + connectionData['source'] + " -> " +
                      connectionData['target'] + ". Input '" + inputName + "' not found on Component:" + targetComponent.getPath())
                connectionFailure = True

            if connectionFailure is False:
                inputPort.setConnection(outputPort, index = connectionData.get('targetIndex', 0))

        Profiler.getInstance().pop()

    def loadRigDefinition(self, jsonData):
        """Load a rig definition from a JSON structure.

        Args:
            jsonData (dict): JSON data containing the rig definition.

        Returns:
            bool: True if successful.

        """

        Profiler.getInstance().push("loadRigDefinition:" + self.getName())

        krakenSystem = KrakenSystem.getInstance()

        if 'name' in jsonData:
            self.setName(jsonData['name'])

        if 'components' in jsonData:
            self._loadComponents(jsonData['components'])

            if 'connections' in jsonData:
                self._makeConnections(jsonData['connections'])

        if 'metaData' in jsonData:

            for k, v in jsonData['metaData'].iteritems():
                self.setMetaDataItem(k, v)

        if 'guideData' in jsonData:
            self.setMetaDataItem('guideData', jsonData['guideData'])

        Profiler.getInstance().pop()

    def writeGuideDefinitionFile(self, filepath):
        """Writes a rig definition to a file on disk.

        Args:
            filepath (str): The file path of the rig definition file.

        Returns:
            bool: True if successful.

        """

        Profiler.getInstance().push("WriteGuideDefinitionFile:" + filepath)

        guideData = self.getRigBuildData()

        # now preprocess the data ready for saving to disk.
        pureJSON = prepareToSave(guideData)

        with open(filepath, 'w') as rigDef:
            rigDef.write(json.dumps(pureJSON, indent=2))

        Profiler.getInstance().pop()

    def getData(self):
        """Get the graph definition of the rig. This method is used to save the state of the guide itself.

        Returns:
            dict: The JSON data struture of the rig data

        """

        guideData = {
            'name': self.getName()
        }

        componentsJson = []
        guideComponents = self.getChildrenByType('Component')
        for component in guideComponents:
            componentsJson.append(component.saveData())
        guideData['components'] = componentsJson

        connectionsJson = []
        for component in guideComponents:
            for i in range(component.getNumInputs()):
                componentInput = component.getInputByIndex(i)
                if componentInput.isConnected():
                    componentOutput = componentInput.getConnection()
                    connectionJson = {
                        'source': componentOutput.getParent().getDecoratedName() + '.' + componentOutput.getName(),
                        'target': component.getDecoratedName() + '.' + componentInput.getName(),
                        'targetIndex': componentInput.getIndex()
                    }
                    connectionsJson.append(connectionJson)

        guideData['connections'] = connectionsJson
        guideData['metaData'] = self._metaData

        return guideData

    def getRigBuildData(self):
        """Get the graph definition of the guide for building the final rig.

        Returns:
            dict: The JSON data struture of the guide rig data

        """

        guideData = self.getData()
        guideJSONData = prepareToSave(guideData)

        rigBuildData = {
            'name': self.getName(),
            'guideData': guideJSONData
        }

        componentsJson = []
        guideComponents = self.getChildrenByType('Component')
        for component in guideComponents:
            componentsJson.append(component.getRigBuildData())

        rigBuildData['components'] = componentsJson

        connectionsJson = []
        for component in guideComponents:
            for i in xrange(component.getNumInputs()):
                componentInput = component.getInputByIndex(i)
                if componentInput.isConnected():
                    componentOutput = componentInput.getConnection()
                    connectionJson = {
                        'source': componentOutput.getParent().getDecoratedName() + '.' + componentOutput.getName(),
                        'target': component.getDecoratedName() + '.' + componentInput.getName(),
                        'targetIndex': componentInput.getIndex()
                    }
                    connectionsJson.append(connectionJson)

        rigBuildData['connections'] = connectionsJson

        return rigBuildData
