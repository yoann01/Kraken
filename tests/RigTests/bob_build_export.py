import json
import os

from kraken import plugins
from kraken.core.objects.rig import Rig
from kraken_examples.bob_guide_data import bob_guide_data
from kraken.core.profiler import Profiler
from kraken.helpers.utility_methods import logHierarchy


Profiler.getInstance().push("bob_build")

bobGuideRig = Rig("char_bob")
bobGuideRig.loadRigDefinition(bob_guide_data)

bobGuideRig.writeRigDefinitionFile('bob_guide.krg')

bobGuideRig2 = Rig()
bobGuideRig2.loadRigDefinitionFile('bob_guide.krg')

logHierarchy(bobGuideRig2)

os.remove('bob_guide.krg')