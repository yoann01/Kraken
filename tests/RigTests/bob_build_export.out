char_bob
char_bob.spine:M
char_bob.controls
char_bob.controls.spine:M
char_bob.controls.spine:M.inputs
char_bob.controls.spine:M.inputs.mainSrt
char_bob.controls.spine:M.outputs
char_bob.controls.spine:M.outputs.cog
char_bob.controls.spine:M.outputs.spineBase
char_bob.controls.spine:M.outputs.pelvis
char_bob.controls.spine:M.outputs.spineEnd
char_bob.controls.spine:M.cogPosition
char_bob.controls.spine:M.spine01Position
char_bob.controls.spine:M.spine02Position
char_bob.controls.spine:M.spine03Position
char_bob.controls.spine:M.spine04Position
char_bob.controls.neck:M
char_bob.controls.neck:M.inputs
char_bob.controls.neck:M.inputs.neckBase
char_bob.controls.neck:M.outputs
char_bob.controls.neck:M.outputs.neck
char_bob.controls.neck:M.outputs.neckEnd
char_bob.controls.neck:M.neck
char_bob.controls.neck:M.neckEnd
char_bob.controls.head:M
char_bob.controls.head:M.inputs
char_bob.controls.head:M.inputs.headBase
char_bob.controls.head:M.outputs
char_bob.controls.head:M.outputs.head
char_bob.controls.head:M.outputs.jaw
char_bob.controls.head:M.outputs.eyeL
char_bob.controls.head:M.outputs.eyeR
char_bob.controls.head:M.head
char_bob.controls.head:M.headEnd
char_bob.controls.head:M.eyeLeft
char_bob.controls.head:M.eyeRight
char_bob.controls.head:M.jaw
char_bob.controls.Clavicle:L
char_bob.controls.Clavicle:L.inputs
char_bob.controls.Clavicle:L.inputs.spineEnd
char_bob.controls.Clavicle:L.outputs
char_bob.controls.Clavicle:L.outputs.clavicle
char_bob.controls.Clavicle:L.outputs.clavicleEnd
char_bob.controls.Clavicle:L.clavicle
char_bob.controls.Clavicle:L.clavicleUpV
char_bob.controls.Clavicle:L.clavicleEnd
char_bob.controls.Clavicle:R
char_bob.controls.Clavicle:R.inputs
char_bob.controls.Clavicle:R.inputs.spineEnd
char_bob.controls.Clavicle:R.outputs
char_bob.controls.Clavicle:R.outputs.clavicle
char_bob.controls.Clavicle:R.outputs.clavicleEnd
char_bob.controls.Clavicle:R.clavicle
char_bob.controls.Clavicle:R.clavicleUpV
char_bob.controls.Clavicle:R.clavicleEnd
char_bob.controls.Arm:L
char_bob.controls.Arm:L.inputs
char_bob.controls.Arm:L.inputs.globalSRT
char_bob.controls.Arm:L.inputs.root
char_bob.controls.Arm:L.outputs
char_bob.controls.Arm:L.outputs.bicep
char_bob.controls.Arm:L.outputs.forearm
char_bob.controls.Arm:L.outputs.wrist
char_bob.controls.Arm:L.bicep
char_bob.controls.Arm:L.forearm
char_bob.controls.Arm:L.wrist
char_bob.controls.Arm:R
char_bob.controls.Arm:R.inputs
char_bob.controls.Arm:R.inputs.globalSRT
char_bob.controls.Arm:R.inputs.root
char_bob.controls.Arm:R.outputs
char_bob.controls.Arm:R.outputs.bicep
char_bob.controls.Arm:R.outputs.forearm
char_bob.controls.Arm:R.outputs.wrist
char_bob.controls.Arm:R.bicep
char_bob.controls.Arm:R.forearm
char_bob.controls.Arm:R.wrist
char_bob.controls.Leg:L
char_bob.controls.Leg:L.inputs
char_bob.controls.Leg:L.inputs.globalSRT
char_bob.controls.Leg:L.inputs.pelvisInput
char_bob.controls.Leg:L.inputs.ikTarget
char_bob.controls.Leg:L.outputs
char_bob.controls.Leg:L.outputs.femur
char_bob.controls.Leg:L.outputs.shin
char_bob.controls.Leg:L.outputs.ikHandle
char_bob.controls.Leg:L.outputs.legEnd
char_bob.controls.Leg:L.outputs.legEndFK
char_bob.controls.Leg:L.femur
char_bob.controls.Leg:L.knee
char_bob.controls.Leg:L.ankle
char_bob.controls.Leg:R
char_bob.controls.Leg:R.inputs
char_bob.controls.Leg:R.inputs.globalSRT
char_bob.controls.Leg:R.inputs.pelvisInput
char_bob.controls.Leg:R.inputs.ikTarget
char_bob.controls.Leg:R.outputs
char_bob.controls.Leg:R.outputs.femur
char_bob.controls.Leg:R.outputs.shin
char_bob.controls.Leg:R.outputs.ikHandle
char_bob.controls.Leg:R.outputs.legEnd
char_bob.controls.Leg:R.outputs.legEndFK
char_bob.controls.Leg:R.femur
char_bob.controls.Leg:R.knee
char_bob.controls.Leg:R.ankle
char_bob.neck:M
char_bob.head:M
char_bob.Clavicle:L
char_bob.Clavicle:R
char_bob.Arm:L
char_bob.Arm:R
char_bob.Leg:L
char_bob.Leg:R
